packages/core/src/components/axes/ruler.ts

```diff
@@ -0,0 +1,187 @@
+// Internal Imports
+import { Component } from "../component";
+import { DOMUtils } from "../../services";
+import { TooltipTypes, ScaleTypes } from "../../interfaces";
+
+// D3 Imports
+import { mouse, Selection } from "d3-selection";
+
+import { isEqual, flatten } from "lodash-es";
+
+type GenericSvgSelection = Selection<SVGElement, any, SVGElement, any>;
+
+const THRESHOLD = 5;
+
+/** check if x is inside threshold area extents  */
+function pointIsMatch(dx: number, x: number) {
+	return dx > x - THRESHOLD && dx < x + THRESHOLD;
+}
+
+export class Ruler extends Component {
+	type = "ruler";
+	backdrop: GenericSvgSelection;
+	hoveredElements: GenericSvgSelection;
+
+	render() {
+		const domainAxisPosition = this.services.cartesianScales.getDomainAxisPosition();
+		const domainScaleType = this.services.cartesianScales.getScaleTypeByPosition(
+			domainAxisPosition
+		);
+		const isTimeSeries = domainScaleType === ScaleTypes.TIME;
+
+		// if scale type is not timeSeries do not show rule
+		if (!isTimeSeries) {
+			return;
+		}
+
+		this.drawBackdrop();
+		this.addBackdropEventListeners();
+	}
+
+	showRuler([x, y]: [number, number]) {
+		const svg = this.parent;
+		const ruler = DOMUtils.appendOrSelect(svg, "g.ruler");
+		const line = DOMUtils.appendOrSelect(ruler, "line.ruler-line");
+		const dataPoints: GenericSvgSelection = svg.selectAll(
+			"[role=graphics-symbol]"
+		);
+		const { datasets } = this.model.getDisplayData();
+		const mainXScale = this.services.cartesianScales.getMainXScale();
+		const mainYScale = this.services.cartesianScales.getMainYScale();
+		const [yScaleEnd, yScaleStart] = mainYScale.range();
+
+		const scaledData: number[] = flatten(datasets.map(({ data }) =>
+			data.map((d, i) => Number(this.services.cartesianScales.getDomainValue(d, i)))
+		));
+
+		/**
+		 * Find matches, reduce is used instead of filter
+		 * to only get elements which belong to the same axis coordinate
+		 */
+		const scaledValuesMatches: number[] = scaledData.reduce((acc, cur) => {
+			const sampleAccValue = acc[0];
```

Comment:
it's a bit unclear to me what the difference between `acc` & `sampleAccValue` is, and why we need to use it. Maybe a more readable variable name + some comments?

Reply:
`acc` here is the accumulator array, basically the "result" of the reducer, `sampleAccValue` is instead a "sample" value taken from the accumulator array, it is used to do the various comparisons with the current value being processed. 
I've added some more comments, hope it's clearer now

Reply:
The reason behind the need of `sampleAccValue` is that we need to do an additional filter to only store data points which belong to the exact same axis coordinate, not just those that make it though the `pointIsWithinThreshold()` condition